src/base/materialized.ts

```diff
@@ -230,84 +231,148 @@ export class MaterializedChatMessage implements IMaterializedNode {
+			.filter(element => typeof element === 'string')
+			.join('').trim();
+
+		if (this.text.some(element => element instanceof MaterializedChatMessageImage)) {
+			if (this.role !== ChatRole.User) {
+				throw new Error('Only User messages can have images');
+			}
+
+			let prompts: ChatCompletionContentPart[] = this.text.map(element => {
+				if (typeof element === 'string') {
+					return { type: 'text', text: element };
+				} else if (element instanceof MaterializedChatMessageImage) {
+					return {
+						type: 'image_url',
+						image_url: { url: getEncodedBase64(element.src), detail: element.detail },
+					};
+				} else {
+					throw new Error('Unexpected element type');
+				}
+			});
+
+			return {
+				role: ChatRole.User,
+				content: prompts,
+			};
+		}
+
 		if (this.role === ChatRole.System) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				...(this.name ? { name: this.name } : {}),
 			};
 		} else if (this.role === ChatRole.Assistant) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				...(this.toolCalls ? { tool_calls: this.toolCalls } : {}),
 				...(this.name ? { name: this.name } : {}),
 			};
 		} else if (this.role === ChatRole.User) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				...(this.name ? { name: this.name } : {}),
 			};
 		} else if (this.role === ChatRole.Tool) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				tool_call_id: this.toolCallId,
 			};
 		} else {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				name: this.name!,
 			};
 		}
 	}
 }
 
+export class MaterializedChatMessageImage implements IMaterializedNode {
+	constructor(
+		public readonly id: number,
+		// public readonly role: ChatRole,
+		public readonly src: string,
+		public readonly priority: number,
+		public readonly metadata: PromptMetadata[] = [],
+		public readonly lineBreakBefore: LineBreakBefore,
+		public readonly detail?: 'low' | 'high',
+	) { }
+	upperBoundTokenCount(tokenizer: ITokenizer): Promise<number> {
+		return this.upperBoundTokenCount(tokenizer);
+	}
+	tokenCount(tokenizer: ITokenizer): Promise<number> {
+		return this._tokenCount(tokenizer);
+	}
+
+	private readonly _tokenCount = once(async (tokenizer: ITokenizer) => {
+		return 0;
+	});
+
+	isEmpty: boolean = false;
+}
+
 function isContainerType(
 	node: MaterializedNode
 ): node is MaterializedContainer | MaterializedChatMessage {
-	return !(node instanceof MaterializedChatMessageTextChunk);
+	return !(node instanceof MaterializedChatMessageTextChunk || node instanceof MaterializedChatMessageImage);
 }
 
 function assertContainerOrChatMessage(
 	v: MaterializedNode
-): asserts v is MaterializedContainer | MaterializedChatMessage {
-	if (!(v instanceof MaterializedContainer) && !(v instanceof MaterializedChatMessage)) {
+): asserts v is MaterializedContainer | MaterializedChatMessage | MaterializedChatMessageImage {
```

Comment:
This shouldn't assert an image I don't think. It should only assert that the node is a chat message or can contain chat message.